### PR TITLE
ROX-20232: Extract common pipelines for tekton pipeline runs

### PR DESCRIPTION
**Pull request by rhybrillou**
## Description

The `PipelineRuns` in the `.tekton` subdirectory have a very similar structure (parameters, results, workspaces, sequence of tasks). This PR aims at reducing the duplication across `pipelineSpec` for the various `PipelineRuns` through the addition of common `Pipeline` objects referenced with appropriate parameters.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

TODO(replace-me)  
Use this space to explain **how you validated** that **your change functions exactly how you expect it**.
Feel free to attach JSON snippets, curl commands, screenshots, etc. Apply a simple benchmark: would the information you
provided convince any reviewer or any external reader that you did enough to validate your change.

It is acceptable to assume trust and keep this section light, e.g. as a bullet-point list.

It is acceptable to skip testing in cases when CI is sufficient, or it's a markdown or code comment change only.  
It is also acceptable to skip testing for changes that are too taxing to test before merging. In such case you are
responsible for the change after it gets merged which includes reverting, fixing, etc. Make sure you validate the change
ASAP after it gets merged or explain in PR when the validation will be performed.  
Explain here why you skipped testing in case you did so.

Have you created automated tests for your change? Explain here which validation activities you did manually and why so.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Review by msugakov — COMMENTED** [review]
Thanks for contributing this @rhybrillou , this is awesome!

Notice however that CI did not run. CI gets triggered when the branch has either `konflux` or `rhtap` in the name. Could you please push to a new one?

I suspect GitHub won't allow changing a branch of _this_ PR and it would force you to create a new one. I'll check and try answer your comments here.

File: .tekton/scanner-v4-db-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    - name: pipelinerun-name
+      value: $(context.pipelineRun.name)
+    - name: git-url
+      value: $(tasks.clone-repository.results.url)?rev=$(tasks.clone-repository.results.commit)
+    - name: image-url
+      value: $(params.output-image)
+    - name: build-task-status
+      value: $(tasks.build-container.status)
+    - name: source-image-url
+      value: $(tasks.build-source-image.results.SOURCE_IMAGE_URL)
```
[the remark this answers is not in the document]
Comment (msugakov) [reply .tekton/scanner-v4-db-pipeline.yaml:32]: You're right, that should be a correct task definition, and it does not have `source-image-url` parameter. We should drop it here too.

Our pipelines were majorly rendered from a [template](https://github.com/redhat-appstudio/build-definitions/blob/main/pipelines/template-build/template-build.yaml). The template gets updates but not the pipelines that were already generated.

File: .tekton/scanner-v4-db-pipeline.yaml
[the remark this answers is not in the document]
Comment (msugakov) [reply .tekton/scanner-v4-db-pipeline.yaml:1]: Yes, to me, it looks to be the same as `basic-component-pipeline.yaml`. Is there anything suspicious that you noticed which may not fit basic component?

File: .tekton/scanner-v4-db-pipeline.yaml [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+  tasks:
+
+  - name: init
+    params:
+    - name: image-url
+      value: $(params.output-image)
+    - name: rebuild
+      value: $(params.rebuild)
+    - name: skip-checks
+      value: $(params.skip-checks)
```
[the remark this answers is not in the document]
Comment (msugakov) [reply .tekton/scanner-v4-db-pipeline.yaml:135]: I think your observations are correct but I did not get the point, sorry. Do you mean that we don't define `skip-checks` at the `spec.params` level and so rely on default value everywhere?

I did not have a case to use `skip-checks` but I suppose it exists for a situation when we need to "just build" without quality gates.

`init` task is very slim. All it does is checks whether the image with the desired tag already exists and whether we want to rebuild it. The subsequent build-container relies on that boolean. I decided that we want to always rebuild the image, therefore set `params.rebuild` to `true`. In this case the value of `skip-checks` does not change anything.

File: .tekton/scanner-v4-db-pipeline.yaml
[the remark this answers is not in the document]
Comment (rhybrillou) [reply .tekton/scanner-v4-db-pipeline.yaml:1]: Nothing suspicious, My development approach was to first extract a pipeline per component, compare with the `-pull-request.yaml` and `-push.yaml` files to see if there was any noticeable difference between the pipeline spec, then compare the extracted pipelines and redirect the components to a common pipeline when applicable. I noticed differences in the definition of the pipeline for `scanner-v4-db` compared to the one used for the other "simpler" components.
My goal is to reduce the number of pipelines as well, in order to reduce the maintenance cost of adding tasks or steps in the pipelines (e.g. change the tag extraction, or add a step to scan the images with our own scanner). So if the build of scanner-v4-db can be fit in the "common" pipeline, that is one pipeline less to maintain.
The diffs I noticed with `basic-component-pipeline.yaml` are the places where I added comments in the file.

File: .tekton/scanner-v4-db-pipeline.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rhybrillou) [reply .tekton/scanner-v4-db-pipeline.yaml:135]: As per comment at file level, I was more investigating the diff in the parameters passed to `init` and their impact as compared to the `basic-component-pipeline.yaml`. In the end, I'm fine with rebuilding the image each time.
I would then add this `skip-checks` variable to the init task parameters in the `basic-component-pipeline.yaml` and use the common pipeline to build `scanner-v4-db`.

File: .tekton/scanner-v4-db-pipeline.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rhybrillou) [reply .tekton/scanner-v4-db-pipeline.yaml:32]: Thanks for pointing me at that template. It helps with the understanding of how things are.

**Comment by msugakov** [comment]
Let's close this out in favor of <https://github.com/stackrox/stackrox/pull/10866>?

**Comment by rhybrillou** [comment]
Closing this PR in favour of #10866

**Pull request closed without merging**